File: DynamicProgramming/lengthOfLIS.js
const lengthOfLISTopDown = (nums) => {
  const memo = new Map();

  const getLongestSubsequence = (i) => {
    if (memo.has(i)) {
      return memo.get(i);
    }

    let longest = 1; // base case

    for (let j = 0; j < i; j++) {
      if (nums[j] < nums[i]) {
        longest = Math.max(getLongestSubsequence(j) + 1, longest);
      }
    }

    memo.set(i, longest);

    return longest;
  }

  let longest = 0;
  for (let i = 0; i < nums.length; i++) {
    longest = Math.max(getLongestSubsequence(i), longest);
  }

  return longest;
}

const nums = [1, 3, 6, 7, 9, 4, 10, 5, 6];
console.log(lengthOfLISTopDown(nums));

const lengthOfLISBottomUp = (nums) => {
  const longestSubsequences = new Array(nums.length).fill(1);
  let longest = 1;
  for (let i = 1; i < nums.length; i++) {
    for (let j = 0; j < i; j++) {
      if (nums[j] < nums[i]) {
        longestSubsequences[i] = Math.max(longestSubsequences[j] + 1, longestSubsequences[i]);
        longest = Math.max(longestSubsequences[i], longest);
      }
    }
  }

  return longest;
}

console.log(lengthOfLISBottomUp(nums));

// Same Time and Space Complexities
// Time O(n^2)
// Space O(n)

const lengthOfLISBinarySearch = (nums) => {
  const positions = new Array(nums.length).fill(0);

  const binarySearch = (target, range) => {
    let left = 0;
    let right = range;
    while (left < right) {
      const mid = Math.floor((left + right) / 2);
      if (target <= positions[mid]) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  let len = 0;
  for (let num of nums) {
    const pos = binarySearch(num, len);
    positions[pos] = num;
    if (pos === len) len++;
  }
  
  return len;
}

console.log(lengthOfLISBinarySearch(nums));

File: DynamicProgramming/LengthOfLIS.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LengthOfLIS {
  // TOP DOWN APPROACH
  Map<Integer, Integer> memo;
  int[] nums;

  public int lengthOfLIS(int[] nums) {
    memo = new HashMap<>();
    this.nums = nums;

    int longest = 1;
    for (int i = 0; i < nums.length; i++) {
      longest = Math.max(getLongest(i), longest);
    }
    return longest;
  }

  private int getLongest(int i) {
    if (memo.containsKey(i)) {
      return memo.get(i);
    }

    int longest = 1;
    for (int j = 0; j < i; j++) {
      if (nums[j] < nums[i]) {
        longest = Math.max(getLongest(j) + 1, longest);
      }
    }

    memo.put(i, longest);
    return longest;
  }

  // BOTTOM UP APPROACH
  public int lengthOfLISBottomUp(int[] nums) {
    int[] dp = new int[nums.length];
    Arrays.fill(dp, 1);
    int longest = 1;
    for (int i = 1; i < nums.length; i++) {
      for (int j = 0; j < i; j++) {
        if (nums[j] < nums[i]) {
          dp[i] = Math.max(dp[j] + 1, dp[i]);
          longest = Math.max(dp[i], longest);
        }
      }
    }

    return longest;
  }

  // BINARY SEARCH APPROACH
  public int lengthOfLISBS(int[] nums) {
    // tails[i] will hold smallest possible tail for
    // subsequence of length i + 1
    List<Integer> tails = new ArrayList<>();
    for (int num : nums) {
      // Find smallest index where num <= tails[i]
      int left = 0;
      int right = tails.size();
      while (left < right) {
        int mid = left + (right - left) / 2;
        if (num <= tails.get(mid)) {
          right = mid;
        } else {
          left = mid + 1;
        }
      }
      if (left == tails.size()) {
        // We have extended the longest subsequence
        tails.add(num);
      } else {
        // We have found a more optimal tail for
        // subsequence of length left + 1
        tails.set(left, num);
      }
    }
    
    return tails.size();
  }
}